finality-aleph/src/sync/handler/mod.rs

```diff
@@ -559,9 +569,175 @@ mod tests {
+
+        assert!(
+            handler.handle_internal_request(&top).expect("should work"),
+            "should be newly required"
+        );
+        assert!(
+            handler.interest_provider().get(&top)
+                == Interest::Required {
+                    know_most: HashSet::new(),
+                    branch_knowledge: LowestId(top.clone()),
+                },
+            "should be required"
+        );
+
+        let (maybe_id, maybe_error) = handler.handle_request_response(
+            branch
+                .iter()
+                .cloned()
+                .rev()
+                .map(ResponseItem::Header)
+                .collect(),
+            peer_id,
+        );
+
+        assert!(
+            maybe_id.is_none(),
+            "should not create new highest justified"
+        );
+        assert!(maybe_error.is_none(), "should work");
+
+        branch
+    }
+
+    fn create_dangling_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        height: BlockNumber,
+        length: usize,
+        peer_id: MockPeerId,
+    ) -> (MockIdentifier, MockIdentifier) {
+        let bottom = MockIdentifier::new_random(height);
+        let top = grow_light_branch(handler, &bottom, length, peer_id)
+            .last()
+            .expect("branch should not be empty")
+            .id();
+        assert_dangling_branch_required(handler, &top, &bottom, HashSet::from_iter(vec![peer_id]));
+        (bottom, top)
+    }
+
+    fn finalize_light_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        branch: Vec<MockHeader>,
+        peer_id: MockPeerId,
+    ) {
+        let blocks: Vec<MockBlock> = branch
+            .iter()
+            .cloned()
+            .map(|header| MockBlock::new(header, true))
+            .collect();
+        let (maybe_id, maybe_error) = handler.handle_request_response(
+            blocks
+                .iter()
+                .cloned()
+                .map(ResponseItem::Block)
+                .chain(
+                    branch
+                        .iter()
+                        .cloned()
+                        .map(MockJustification::for_header)
+                        .map(ResponseItem::Justification),
+                )
+                .collect(),
+            peer_id,
+        );
+        let new_top = branch.last().expect("branch should not be empty").id();
+        assert!(
+            maybe_id == Some(new_top),
+            "should create new highest justified"
+        );
+        assert!(maybe_error.is_none(), "should work");
+    }
+
+    #[test]
+    fn creates_dangling_branch() {
```

Comment:
what does it tests?

Reply:
mostly the bug that I'm fixing

the idea is that `create_dangling_branch` will be a part of other more complex tests, so I want to have a separate test just for it in case it breaks, and then we won't have to wonder which one of 10 parts of a large test doesn't work

Reply:
testing the test code, nice <#